### ISSUE TITLE
End of file reached while trying to read queue item

### ISSUE THREAD

**Issue by chuckchopp** [i-e-b/DiskQueue#13]
I am using DiskQueue v1.3.0 in a c# project on Windows Server [Win2K12R2, Win2K16, Win2K19] that runs as a native NT service.  The interactions with the queue take place within the service's process, with one thread being the event provider and another thread being the event consumer.  The event provider detects events of interest occurring on the server and places them in the on-disk queue as rapidly as possible while the event consumer is responsible for streaming events out to the network.  All Enqueue() and Dequeue() method calls are immediately followed by a Flush() call out of a sense of paranoia about ensuring that the operations are committed to disk immediately.

After a couple of days [2 to 4] of continuous execution, I find that dequeue operations start to fail with the following exception being caught & logged:

System.InvalidOperationException: End of file reached while trying to read queue item
   at DiskQueue.Implementation.PersistentQueueImpl.ReadEntriesFromFile(Entry firstEntry, Int64 currentBufferSize)
   at DiskQueue.Implementation.PersistentQueueImpl.ReadAhead()
   at DiskQueue.Implementation.PersistentQueueImpl.Dequeue()
   at DiskQueue.Implementation.PersistentQueueSession.Dequeue()

The "while" loop in my code that dequeues events from the on-disk queue catches the exception, logs it, sleeps for 100ms and then iterates through the loop again.  Once this problem starts occurring, the condition does not clear itself and my log files fill up with messages showing the same exception occurring repeatedly.

At the time that this happens, I find that the directory used by DiskQueue contains the following:

02/03/2021  06:45 AM    <DIR>          .
02/03/2021  06:45 AM    <DIR>          ..
02/03/2021  12:13 AM                 0 data.158
02/03/2021  03:46 AM        33,557,302 data.159
02/03/2021  06:45 AM        33,557,302 data.160
01/31/2021  07:28 PM                 0 lock
02/03/2021  06:45 AM                12 meta.state
02/03/2021  06:45 AM            96,834 transaction.log
               6 File(s)     67,211,450 bytes
               2 Dir(s)   7,109,398,528 bytes free

Nothing appears unusual in any way w/respect to the number of data files, the sizes of the data files or the size of the transaction log.

Are there any known issues with DiskQueue that would cause the Dequeue() method to attempt to read data for an event/message which appears to be present based on metadata but which was never actually committed to the data files on disk?

Is there any recommendation regarding a fallback/recovery strategy short of disposing of the DiskQueue class instance and then instantiating a new one?

Is there any test that can be performed to verify that the on-disk queue files are corrupted in some way so that they can be deleted?

**Comment by i-e-b** [comment]
Hi. Thanks for the bug report.

At first glance, it looks like something has gone wrong in the underlying file system -- but DiskQueue certainly shouldn't get itself in an endless loop.

I'll take a look and report back.

**Comment by i-e-b** [comment]
Please try https://www.nuget.org/packages/DiskQueue/1.3.1

This has two changes
1. All flushes should correctly call the operating system to flush to disk
2. Adds a setting `queue.Internals.AllowTruncatedEntries = true;` which will skip the error.

Please try the 1.3.1 version *without* the truncated entries setting, and let me know what happens.

Thanks.

**Comment by chuckchopp** [comment]
Thanks!

I will give it a try here as soon as I get some in-progress code edits
completed and can integrate the new build of DiskQueue into my project.

I think that I have a better understanding of how things got corrupted to
start with.  It looks like an unrelated unhandled exception occurred that
terminated my application's service process while it was actively
performing enqueue operations to the persistent queue.  The service was
configured to automatically restart, so the Windows Service Control Manager
diligently started it again, and after that point, all dequeue operations
caused the exception that I reported.  However, it appears that enqueue
operations could still be performed successfully, with the result being
that my application was continuing to add entries to the persistent queue
but was completely unable to get them out of the queue due to the exception
occurring during the dequeue operation.

With that behavior in mind, how does that affect your recommendation to
test with AllowTruncatedEntires set to false?  Is that recommendation just
for testing the changes to see if they are effective at preventing the
exception in the dequeue operation?  From what I was observing, it seems
like I may well need that setting enabled to get past the corrupted portion
of the queue.

Also, having considered that such a problem could occur again in the future
simply by having a roque administrator forcefully terminating my
application's service process or the O.S. itself crashing, I implemented
code to simply delete all files from the directory used by the persistent
queue before instantiating the persistent queue class.  This code only
executes if a specific command line parameter is passed to my application
at start up time.  I wanted to know if there was any latent support for
this within DiskQueue.PersistentQueue w/respect to c'tor method overloads
or possibly a static class method that would effectively obliterate the
queue content and create a new empty queue.


Regards,

Chuck
